### PR TITLE
Fix validation layer errors

### PR DESCRIPTION
**Pull request by zakorgy**
The first commit fixes the first error in https://github.com/szeged/webrender/issues/80#issuecomment-488642315, the second one fixes the second error and the errors related to the document example.

The WIP is there because I still need to figure out the dx12 debug layer errors and these changes are not tested on macOs yet.

### CODE REVIEW

**Comment by zakorgy** [comment]
A quick update on this:
- The gfx update fixes https://github.com/szeged/webrender/issues/277 as we expected.
- Most of the validation errors are fixed from https://github.com/szeged/webrender/issues/80#issuecomment-488642315 (both Vulkan and DX12) There is two more left, which are related to the window resizing, but we can take care of those later since  those only show up for one frame after the resize, hence we can threat them as low priority issues.

File: webrender/src/device/gfx/device.rs [hunk 1]
```diff
@@ -818,6 +827,8 @@ impl<B: hal::Backend> Device<B> {
             Backbuffer::Framebuffer(fbo) => (vec![], vec![fbo], vec![]),
         };
 
+        println!("Frame images: {:?}\nFrame depths: {:?}", frame_images, frame_depths);
```
Comment (kvark) [inline webrender/src/device/gfx/device.rs]: let's use `log` crate macros?

File: webrender/src/device/gfx/device.rs [hunk 2]
```diff
@@ -1249,31 +1259,31 @@ impl<B: hal::Backend> Device<B> {
...
             program.bind_texture(
                 &self.device,
                 desc_set,
                 &self.images[&self.bound_textures[index]].core,
                 sampler_name,
+                &mut command_buffer,
             );
         }
-        let desc_set = self.descriptor_pools_global.get(&program.shader_kind);
+        let desc_set = self.descriptor_pools_global[self.next_id].get(&program.shader_kind);
```
Comment (kvark) [inline webrender/src/device/gfx/device.rs]: if we cache the descriptor sets between frames, and treat each desc set as immutable for as long as it's tracked, we don't need separate pools per frame. So `descriptor_pools_xxx` would need to switch from being a small vector of at least 1 element to exactly a single instance of a descriptor pool.

File: webrender/src/device/gfx/device.rs [hunk 3]
```diff
@@ -1341,41 +1351,95 @@ impl<B: hal::Backend> Device<B> {
...
                     &self.framebuffers[self.current_frame_id],
                     self.surface_format,
+                    (None, false)
                 )
             } else {
                 (
+                    &self.frame_images[self.current_frame_id],
                     &self.framebuffers_depth[self.current_frame_id],
                     self.surface_format,
+                    (Some(&self.frame_depths[self.current_frame_id].core), false)
```
Comment (kvark) [inline webrender/src/device/gfx/device.rs]: let's move this out as `let depth_image = match self.current_depth_test {..};`

File: webrender/src/device/gfx/device.rs [hunk 4]
```diff
@@ -1341,41 +1351,95 @@ impl<B: hal::Backend> Device<B> {
...
+                before_depth_state = Some(depth_img.state.get());
+                if let Some(barrier) = depth_img.transit(
+                    hal::image::Access::empty(),
+                    hal::image::Layout::DepthStencilAttachmentOptimal,
+                    depth_img.subresource_range.clone(),
+                   None,
+                ) {
+                    cmd_buffer.pipeline_barrier(
+                        PipelineStage::EARLY_FRAGMENT_TESTS
+                            .. PipelineStage::LATE_FRAGMENT_TESTS,
```
Comment (kvark) [inline webrender/src/device/gfx/device.rs]: I think these barriers are incorrect. When you are saying that this depth image needs to get DepthStencilAttachmentOptimal layout, you want this to happen before this image starts to be used as a d/s attachment. So your barrier should end at EARLY_FRAGMENT_TESTS (and start point would depend on what attachment/use it was in, previously)

File: webrender/src/device/gfx/device.rs (hunk 2 above)
Comment (zakorgy) [reply webrender/src/device/gfx/device.rs]: After this PR and https://github.com/szeged/webrender/pull/282 lands we can move forward in this direction.

**Pull request merged**